### PR TITLE
Add Wi4MPI as an MPI provider

### PR DESCRIPTION
**Pull request by spoutn1k**
Hello all ! I needed to compile a package with the MPI interface Wi4MPI provides, so I whipped this up to allow to specify Wi4MPI as provider.

### CODE REVIEW

**Review by adrien-cotte — APPROVED** [review]
Great job 😄👌

**Review by marc-joos — APPROVED** [review]
Thank you for the good work!

**Review by alalazo — CHANGES_REQUESTED** [review]
Depending on a virtual and providing the same virtual is not supported by our software model. Can you concretize anything with `wi4mpi@3.6:`?

**Comment by spoutn1k** [comment]
Yep, apart from some linking errors I was able to compile code with this package.

```
root@4a9f8221f297:/# spack spec petsc@3.19.0+rocm amdgpu_target=gfx90a ^wi4mpi ^openblas
Input spec
--------------------------------
petsc@3.19.0+rocm amdgpu_target=gfx90a
    ^openblas
    ^wi4mpi

Concretized
--------------------------------
petsc@3.19.0%gcc@11.1.0~X~batch~cgns~complex~cuda~debug+double~exodusii~fftw+fortran~giflib+hdf5~hpddm~hwloc+hypre~int64~jpeg~knl~kokkos~libpng~libyaml~memkind+metis~mkl-pardiso~mmg~moab~mpfr+mpi~mumps~openmp~p4est~parmmg~ptscotch~random123+rocm~saws~scalapack+shared~strumpack~suite-sparse+superlu-dist~tetgen~trilinos~valgrind amdgpu_target=gfx90a build_system=generic clanguage=C arch=linux-ubuntu20.04-x86_64
    ^diffutils@3.8%gcc@11.1.0 build_system=autotools arch=linux-ubuntu20.04-x86_64
        ^libiconv@1.17%gcc@11.1.0 build_system=autotools libs=shared,static arch=linux-ubuntu20.04-x86_64
    ^hdf5@1.14.0%gcc@11.1.0~cxx~fortran~hl~ipo~java+mpi+shared~szip~threadsafe+tools api=default build_system=cmake build_type=RelWithDebInfo patches=0b5dd6f arch=linux-ubuntu20.04-x86_64
        ^cmake@3.25.2%gcc@11.1.0~doc+ncurses+ownlibs~qt build_system=generic build_type=Release arch=linux-ubuntu20.04-x86_64
        ^pkgconf@1.8.0%gcc@11.1.0 build_system=autotools arch=linux-ubuntu20.04-x86_64
    ^hip@5.4.3%gcc@11.1.0~cuda~ipo+rocm build_system=cmake build_type=Release patches=ca523f1 arch=linux-ubuntu20.04-x86_64
        ^comgr@5.4.3%gcc@11.1.0~ipo build_system=cmake build_type=Release arch=linux-ubuntu20.04-x86_64
        ^glx@1.4%gcc@11.1.0 build_system=bundle arch=linux-ubuntu20.04-x86_64
            ^mesa@22.1.6%gcc@11.1.0+glx+llvm+opengl~opengles+osmesa~strip build_system=meson buildtype=release default_library=shared patches=3b8acf5,ee737d1 arch=linux-ubuntu20.04-x86_64
                ^bison@3.8.2%gcc@11.1.0 build_system=autotools arch=linux-ubuntu20.04-x86_64
                ^flex@2.6.3%gcc@11.1.0+lex~nls build_system=autotools arch=linux-ubuntu20.04-x86_64
                    ^findutils@4.9.0%gcc@11.1.0 build_system=autotools patches=440b954 arch=linux-ubuntu20.04-x86_64
                ^glproto@1.4.17%gcc@11.1.0 build_system=autotools arch=linux-ubuntu20.04-x86_64
                    ^util-macros@1.19.3%gcc@11.1.0 build_system=autotools arch=linux-ubuntu20.04-x86_64
                ^libunwind@1.6.2%gcc@11.1.0~block_signals~conservative_checks~cxx_exceptions~debug~debug_frame+docs+pic+tests+weak_backtrace+xz~zlib build_system=autotools components=none libs=shared,static arch=linux-ubuntu20.04-x86_64
                ^libx11@1.7.0%gcc@11.1.0 build_system=autotools arch=linux-ubuntu20.04-x86_64
                    ^inputproto@2.3.2%gcc@11.1.0 build_system=autotools arch=linux-ubuntu20.04-x86_64
                    ^kbproto@1.0.7%gcc@11.1.0 build_system=autotools arch=linux-ubuntu20.04-x86_64
                    ^xextproto@7.3.0%gcc@11.1.0 build_system=autotools arch=linux-ubuntu20.04-x86_64
                    ^xproto@7.0.31%gcc@11.1.0 build_system=autotools arch=linux-ubuntu20.04-x86_64
                    ^xtrans@1.3.5%gcc@11.1.0 build_system=autotools arch=linux-ubuntu20.04-x86_64
                ^libxcb@1.14%gcc@11.1.0 build_system=autotools arch=linux-ubuntu20.04-x86_64
                    ^libxau@1.0.8%gcc@11.1.0 build_system=autotools arch=linux-ubuntu20.04-x86_64
                    ^libxdmcp@1.1.2%gcc@11.1.0 build_system=autotools arch=linux-ubuntu20.04-x86_64
                    ^xcb-proto@1.14.1%gcc@11.1.0 build_system=autotools arch=linux-ubuntu20.04-x86_64
                ^libxext@1.3.3%gcc@11.1.0 build_system=autotools arch=linux-ubuntu20.04-x86_64
                ^libxt@1.1.5%gcc@11.1.0 build_system=autotools arch=linux-ubuntu20.04-x86_64
                    ^libice@1.0.9%gcc@11.1.0 build_system=autotools arch=linux-ubuntu20.04-x86_64
                    ^libsm@1.2.3%gcc@11.1.0 build_system=autotools arch=linux-ubuntu20.04-x86_64
                ^meson@1.0.0%gcc@11.1.0 build_system=python_pip patches=0f0b1bd arch=linux-ubuntu20.04-x86_64
                ^py-mako@1.1.6%gcc@11.1.0 build_system=python_pip arch=linux-ubuntu20.04-x86_64
                    ^py-markupsafe@2.1.1%gcc@11.1.0 build_system=python_pip arch=linux-ubuntu20.04-x86_64
                ^xrandr@1.5.0%gcc@11.1.0 build_system=autotools arch=linux-ubuntu20.04-x86_64
                    ^libxrandr@1.5.0%gcc@11.1.0 build_system=autotools arch=linux-ubuntu20.04-x86_64
                        ^renderproto@0.11.1%gcc@11.1.0 build_system=autotools arch=linux-ubuntu20.04-x86_64
                    ^libxrender@0.9.10%gcc@11.1.0 build_system=autotools arch=linux-ubuntu20.04-x86_64
                    ^randrproto@1.5.0%gcc@11.1.0 build_system=autotools arch=linux-ubuntu20.04-x86_64
        ^hsakmt-roct@5.4.3%gcc@11.1.0~ipo+shared build_system=cmake build_type=Release patches=114b05a arch=linux-ubuntu20.04-x86_64
            ^libdrm@2.4.114%gcc@11.1.0~docs build_system=generic arch=linux-ubuntu20.04-x86_64
                ^libpciaccess@0.16%gcc@11.1.0 build_system=autotools arch=linux-ubuntu20.04-x86_64
                ^libpthread-stubs@0.4%gcc@11.1.0 build_system=autotools arch=linux-ubuntu20.04-x86_64
        ^numactl@2.0.14%gcc@11.1.0 build_system=autotools patches=4e1d78c,62fc8a8,ff37630 arch=linux-ubuntu20.04-x86_64
            ^autoconf@2.69%gcc@11.1.0 build_system=autotools patches=35c4492,7793209,a49dd5b arch=linux-ubuntu20.04-x86_64
            ^automake@1.16.5%gcc@11.1.0 build_system=autotools arch=linux-ubuntu20.04-x86_64
            ^libtool@2.4.7%gcc@11.1.0 build_system=autotools arch=linux-ubuntu20.04-x86_64
            ^m4@1.4.19%gcc@11.1.0+sigsegv build_system=autotools patches=9dc5fbd,bfdffa7 arch=linux-ubuntu20.04-x86_64
                ^libsigsegv@2.13%gcc@11.1.0 build_system=autotools arch=linux-ubuntu20.04-x86_64
        ^perl@5.36.0%gcc@11.1.0+cpanm+open+shared+threads build_system=generic arch=linux-ubuntu20.04-x86_64
            ^berkeley-db@18.1.40%gcc@11.1.0+cxx~docs+stl build_system=autotools patches=26090f4,b231fcc arch=linux-ubuntu20.04-x86_64
        ^perl-file-which@1.22%gcc@11.1.0 build_system=perl arch=linux-ubuntu20.04-x86_64
        ^perl-uri-encode@1.1.1%gcc@11.1.0 build_system=perl arch=linux-ubuntu20.04-x86_64
            ^perl-module-build@0.4224%gcc@11.1.0 build_system=perl arch=linux-ubuntu20.04-x86_64
        ^py-cppheaderparser@2.7.4%gcc@11.1.0 build_system=python_pip arch=linux-ubuntu20.04-x86_64
            ^py-ply@3.11%gcc@11.1.0 build_system=python_pip arch=linux-ubuntu20.04-x86_64
            ^py-setuptools@65.5.0%gcc@11.1.0 build_system=generic arch=linux-ubuntu20.04-x86_64
        ^rocminfo@5.4.3%gcc@11.1.0~ipo build_system=cmake build_type=Release arch=linux-ubuntu20.04-x86_64
        ^roctracer-dev-api@5.4.3%gcc@11.1.0 build_system=generic arch=linux-ubuntu20.04-x86_64
    ^hipblas@5.4.3%gcc@11.1.0~ipo build_system=cmake build_type=Release arch=linux-ubuntu20.04-x86_64
        ^rocm-cmake@5.4.3%gcc@11.1.0~ipo build_system=cmake build_type=Release arch=linux-ubuntu20.04-x86_64
    ^hipsolver@5.4.3%gcc@11.1.0~ipo build_system=cmake build_type=Release arch=linux-ubuntu20.04-x86_64
    ^hipsparse@5.4.3%gcc@11.1.0~ipo build_system=cmake build_type=Release arch=linux-ubuntu20.04-x86_64
        ^git@2.39.1%gcc@11.1.0+man+nls+perl+subtree~svn~tcltk build_system=autotools arch=linux-ubuntu20.04-x86_64
            ^curl@7.87.0%gcc@11.1.0~gssapi~ldap~libidn2~librtmp~libssh~libssh2~nghttp2 build_system=autotools libs=shared,static tls=openssl arch=linux-ubuntu20.04-x86_64
            ^libidn2@2.3.4%gcc@11.1.0 build_system=autotools arch=linux-ubuntu20.04-x86_64
                ^libunistring@1.1%gcc@11.1.0 build_system=autotools arch=linux-ubuntu20.04-x86_64
            ^openssh@9.1p1%gcc@11.1.0+gssapi build_system=autotools arch=linux-ubuntu20.04-x86_64
                ^krb5@1.20.1%gcc@11.1.0+shared build_system=autotools arch=linux-ubuntu20.04-x86_64
                ^libedit@3.1-20210216%gcc@11.1.0 build_system=autotools arch=linux-ubuntu20.04-x86_64
            ^pcre2@10.42%gcc@11.1.0~jit+multibyte build_system=autotools arch=linux-ubuntu20.04-x86_64
    ^hsa-rocr-dev@5.4.3%gcc@11.1.0+image~ipo+shared build_system=cmake build_type=Release patches=71e6851 arch=linux-ubuntu20.04-x86_64
        ^elfutils@0.188%gcc@11.1.0+bzip2~debuginfod~nls+xz~zstd build_system=autotools arch=linux-ubuntu20.04-x86_64
        ^xxd-standalone@8.2.1201%gcc@11.1.0 build_system=makefile arch=linux-ubuntu20.04-x86_64
    ^hypre@2.27.0%gcc@11.1.0~complex~cuda~debug+fortran~gptune~int64~internal-superlu~mixedint+mpi~openmp~rocm+shared~superlu-dist~sycl~umpire~unified-memory build_system=autotools arch=linux-ubuntu20.04-x86_64
    ^llvm-amdgpu@5.4.3%gcc@11.1.0~ipo~link_llvm_dylib~llvm_dylib~openmp+rocm-device-libs build_system=cmake build_type=Release patches=a08bbe1 arch=linux-ubuntu20.04-x86_64
        ^ncurses@6.3%gcc@11.1.0~symlinks+termlib abi=none build_system=autotools arch=linux-ubuntu20.04-x86_64
        ^ninja@1.11.1%gcc@11.1.0+re2c build_system=generic arch=linux-ubuntu20.04-x86_64
            ^re2c@2.2%gcc@11.1.0 build_system=generic arch=linux-ubuntu20.04-x86_64
        ^z3@4.11.2%gcc@11.1.0~gmp~ipo~python build_system=cmake build_type=RelWithDebInfo arch=linux-ubuntu20.04-x86_64
    ^metis@5.1.0%gcc@11.1.0~gdb~int64~ipo~real64+shared build_system=cmake build_type=RelWithDebInfo patches=4991da9,93a7903,b1225da arch=linux-ubuntu20.04-x86_64
    ^openblas@0.3.21%gcc@11.1.0~bignuma~consistent_fpcsr+fortran~ilp64+locking+pic+shared build_system=makefile patches=d3d9b15 symbol_suffix=none threads=openmp arch=linux-ubuntu20.04-x86_64
    ^parmetis@4.0.3%gcc@11.1.0~gdb~int64~ipo+shared build_system=cmake build_type=RelWithDebInfo patches=4f89253,50ed208,704b84f arch=linux-ubuntu20.04-x86_64
    ^python@3.7.15%gcc@11.1.0+bz2+crypt+ctypes+dbm~debug+libxml2+lzma~nis~optimizations+pic+pyexpat+pythoncmd+readline+shared+sqlite3+ssl~tkinter+uuid+zlib build_system=generic patches=0d98e93,f2fd060 arch=linux-ubuntu20.04-x86_64
        ^bzip2@1.0.8%gcc@11.1.0~debug~pic+shared build_system=generic arch=linux-ubuntu20.04-x86_64
        ^expat@2.5.0%gcc@11.1.0+libbsd build_system=autotools arch=linux-ubuntu20.04-x86_64
            ^libbsd@0.11.7%gcc@11.1.0 build_system=autotools arch=linux-ubuntu20.04-x86_64
                ^libmd@1.0.4%gcc@11.1.0 build_system=autotools arch=linux-ubuntu20.04-x86_64
        ^gdbm@1.23%gcc@11.1.0 build_system=autotools arch=linux-ubuntu20.04-x86_64
        ^gettext@0.21.1%gcc@11.1.0+bzip2+curses+git~libunistring+libxml2+tar+xz build_system=autotools arch=linux-ubuntu20.04-x86_64
            ^libxml2@2.10.3%gcc@11.1.0~python build_system=autotools arch=linux-ubuntu20.04-x86_64
            ^tar@1.34%gcc@11.1.0 build_system=autotools zip=pigz arch=linux-ubuntu20.04-x86_64
                ^pigz@2.7%gcc@11.1.0 build_system=makefile arch=linux-ubuntu20.04-x86_64
                ^zstd@1.5.2%gcc@11.1.0+programs build_system=makefile compression=none libs=shared,static arch=linux-ubuntu20.04-x86_64
        ^libffi@3.4.4%gcc@11.1.0 build_system=autotools arch=linux-ubuntu20.04-x86_64
        ^libxcrypt@4.4.33%gcc@11.1.0~obsolete_api build_system=autotools arch=linux-ubuntu20.04-x86_64
        ^openssl@1.1.1s%gcc@11.1.0~docs~shared build_system=generic certs=mozilla arch=linux-ubuntu20.04-x86_64
            ^ca-certificates-mozilla@2023-01-10%gcc@11.1.0 build_system=generic arch=linux-ubuntu20.04-x86_64
        ^readline@8.2%gcc@11.1.0 build_system=autotools patches=bbf97f1 arch=linux-ubuntu20.04-x86_64
        ^sqlite@3.40.1%gcc@11.1.0+column_metadata+dynamic_extensions+fts~functions+rtree build_system=autotools arch=linux-ubuntu20.04-x86_64
        ^util-linux-uuid@2.38.1%gcc@11.1.0 build_system=autotools arch=linux-ubuntu20.04-x86_64
        ^xz@5.4.1%gcc@11.1.0+pic build_system=autotools libs=shared,static arch=linux-ubuntu20.04-x86_64
    ^rocblas@5.4.3%gcc@11.1.0~ipo+tensile amdgpu_target=auto build_system=cmake build_type=Release patches=81591d9 arch=linux-ubuntu20.04-x86_64
        ^msgpack-c@3.1.1%gcc@11.1.0~ipo build_system=cmake build_type=RelWithDebInfo arch=linux-ubuntu20.04-x86_64
        ^py-msgpack@1.0.4%gcc@11.1.0 build_system=python_pip arch=linux-ubuntu20.04-x86_64
            ^py-cython@0.29.32%gcc@11.1.0 build_system=python_pip arch=linux-ubuntu20.04-x86_64
        ^py-pip@23.0%gcc@11.1.0 build_system=generic arch=linux-ubuntu20.04-x86_64
        ^py-pyyaml@6.0%gcc@11.1.0+libyaml build_system=python_pip arch=linux-ubuntu20.04-x86_64
            ^libyaml@0.2.5%gcc@11.1.0 build_system=autotools arch=linux-ubuntu20.04-x86_64
        ^py-virtualenv@16.7.6%gcc@11.1.0 build_system=python_pip arch=linux-ubuntu20.04-x86_64
        ^py-wheel@0.37.1%gcc@11.1.0 build_system=generic arch=linux-ubuntu20.04-x86_64
    ^rocprim@5.4.3%gcc@11.1.0~ipo amdgpu_target=auto build_system=cmake build_type=Release arch=linux-ubuntu20.04-x86_64
    ^rocrand@5.4.3%gcc@11.1.0~ipo amdgpu_target=auto build_system=cmake build_type=Release patches=a35e689 arch=linux-ubuntu20.04-x86_64
    ^rocsolver@5.4.3%gcc@11.1.0~ipo+optimal amdgpu_target=auto build_system=cmake build_type=Release patches=8067bfb arch=linux-ubuntu20.04-x86_64
        ^fmt@9.1.0%gcc@11.1.0~ipo+pic~shared build_system=cmake build_type=RelWithDebInfo cxxstd=11 patches=08fb707 arch=linux-ubuntu20.04-x86_64
    ^rocsparse@5.4.3%gcc@11.1.0~ipo~test amdgpu_target=auto build_system=cmake build_type=Release arch=linux-ubuntu20.04-x86_64
    ^rocthrust@5.4.3%gcc@11.1.0~ipo amdgpu_target=auto build_system=cmake build_type=Release arch=linux-ubuntu20.04-x86_64
    ^superlu-dist@8.1.2%gcc@11.1.0~cuda~int64~ipo~openmp~rocm+shared build_system=cmake build_type=RelWithDebInfo arch=linux-ubuntu20.04-x86_64
    ^wi4mpi@3.6.4%gcc@11.1.0~ipo build_system=cmake build_type=Release arch=linux-ubuntu20.04-x86_64
    ^zlib@1.2.13%gcc@11.1.0+optimize+pic+shared build_system=makefile arch=linux-ubuntu20.04-x86_64
```

**Review by bernhardkaindl — COMMENTED** [review]
@spoutn1k @marcjoos-cea @adrien-cotte:

alalazo blocked merging this by asking this question with the review-request:

> Depending on a virtual and providing the same virtual is not supported by our software model. Can you concretize anything with wi4mpi@3.6:?

It looks like he referred to this:
```py
     depends_on("mpi", when="@:3.5")
+    provides("mpi@:4.0", when="@3.6:")
```

Can you change this part or make comment on it so alalzo can reconsider started review?

**Comment by nhanford** [comment]
Hello @alalazo and @bernhardkaindl ,

Older versions (before 3.6) of WI4MPI depended on a reference implementation of MPI and therefore could not provide the MPI virtual. However, providing the MPI virtual is important for being able to use WI4MPI in its "interface" mode, where the user builds against WI4MPI as a shim implementation and selects which MPI implementation is used at runtime.
I'm wondering what your thoughts are on how we would go forward with this? Should it be split into 2 packages? Should older versions just be dropped from Spack with a note that this provider issue was present?
The other consideration that needs to be taken into account is that we probably need to avoid having the concretizer resolve to WI4MPI unless the user specifies it, as WI4MPI requires *some* MPI implementation to be specified on the system, and therefore should only be used intentionally by users who know this extra step will be required.

Thanks,
Nate

**Comment by marc-joos** [comment]
Hello @alalazo, @bernhardkaindl and @nhanford,
Dropping older versions would be acceptable on our side as the changes between versions are mostly bug fixes and there is no real good reason, as far as I am aware of, to stick to a particular (and old) version.

**Comment by alalazo** [comment]
A few notes:
1. I think I missed at the time that `depends_on` and `provides` are conditional on non-overlapping conditions, so that should be fine
2. On the other hand, I don't see how this package can be used for v3.5 if it is not an mpi provider
3. Even if it is an `mpi` provider, I assume this package needs at least a `run` dependency on another `mpi`, right?

TL;DR: the current directives seem fine, though I don't know if they model what the package does correctly

**Comment by nhanford** [comment]
To address @alalazo's notes:

2. I forgot to mention that there is also a "preload" mode where you can build an application with an MPI implementation of your choice and it will translate to another implementation.
3. That is correct. However, would requiring MPI as a run dependency conflict with the fact that it is a provider of the MPI virtual?

**Comment by alalazo** [comment]
> However, would requiring MPI as a run dependency conflict with the fact that it is a provider of the MPI virtual?

If the dependency is expressed as:
```python
depends_on("mpi", type="run")
```
it is an issue right now, since we unify link/run dependencies, and a package cannot provide `mpi` and depend on it. If the dependency is like:
```python
depends_on("openmpi")
```
without mentioning the virtual, I think it could work. In any case, I'd be happy to test drive some mid-size use case, and see if we need anything to support this middle layer better.

**Comment by spoutn1k** [comment]
The thing is, it does not depend on openMPI to run, but on any installed MPI implementation. This is not a detail, this is the point of Wi4MPI, to create a MPI compatibillity layer.

**Comment by alalazo** [comment]
@nhanford Any action we can take to push this one forward?

**Comment by nhanford** [comment]
Sorry this goes to the edge of my packaging knowledge, but I think your proposal:
```python
depends_on("openmpi")
```
would make sense if we could allow the user to somehow specify *which* MPI(s) they want to use as they need at least one "target" implementation.

**Comment by spoutn1k** [comment]
With recent radical life changes I had no time to look at this. What is the final consensus ?

The following:
```py
depends_on("openmpi")
```
Seems a little restrictive when the point of the project is to be MPI-fluid. Is there any way we could hide those `depends_on` definitions behind mutually-exclusive features flags ? Akin to `wi4mpi@x.y+mpich` or `wi4mpi@z.a+intelmpi` ?

**Comment by haampie** [comment]
Hi @spoutn1k, this package-related pull request can be migrated from `spack/spack` to `spack/spack-packages` using the [**migrate-pkg-prs utility**](https://github.com/spack/migrate-package-prs).

We encourage authors to run the migration script themselves to preserve author attribution.

You can migrate all your open pull requests at once, following the steps in the [documentation](https://github.com/spack/migrate-package-prs?tab=readme-ov-file#usage-instructions).

---

*This comment was posted automatically.*
